### ISSUE TITLE
console broke

### ISSUE THREAD

**Issue by gmt** [Afforess/Factorio-Stdlib#109]
The console doesn't work anymore as documented and it's a bit tricky to salvage the approach of returning a function to require given that require is now off-limits during events (boo! hiss!)

For my own temporary mod hacking needs, I was able to salvage most of the code by insta-global invocation of remote.add_interface but it's not nearly as sexy as your way (be advised that below I disabled some of your features that were diagnostically inconveniencing me):

```-- debug console

local Event = require("stdlib/event/event")
local Game = require("stdlib/game")
local Gui = require("stdlib/event/gui")

-- to activate:
-- /c remote.call("debug_console", "show", game.player)

remote.add_interface("debug_console", {show = function(player)
	local prefix = "gmt"
	local names = {
	    frame = prefix..'_console',
	    scroll = prefix..'_console_scroll',
	    textbox = prefix..'_console_line',
	    enter = prefix..'_console_enter',
	    clear = prefix..'_console_clear',
	    close = prefix..'_console_close',
	}
	if player.gui.left[names.frame] then player.gui.left[names.frame].destroy() end

	local c = player.gui.left.add{type='frame', name = names.frame, direction = 'horizontal'}

	local scroll = c.add{type = 'scroll-pane', name = names.scroll}
	scroll.style.minimal_width=600
	scroll.style.maximal_width=600
	scroll.style.maximal_height=150
	scroll.style.minimal_height=150

	local t = scroll.add{type = 'text-box', name = names.textbox}
	t.style.minimal_width=600
	t.style.maximal_width=600
	t.style.minimal_height=150

	c.add{type = 'button', name = names.enter, caption = '<', tooltip = "Run Script"}
	c.add{type = 'button', name = names.clear, caption = 'C', tooltip = "Clear Input"}
	c.add{type = 'button', name = names.close, caption = "X", tooltip = "Close"}
	
	local function enter(event)
		local p = Game.get_player(event.player_index)
		local s = event.element.parent[names.scroll][names.textbox].text

		local ok, err = pcall(function() return loadstring(s)() end )
		if not ok then print(err) end
		--pcall(loadstring("return function(a, b, c) ".. chunkstring .. " end"))
		--game.write_file(prefix..'/console.log', s..'\n', true, p.index)
	end
	Gui.on_click('^'..names.enter..'$', enter)

	local function close(event)
		event.element.parent.destroy()
	end
	Gui.on_click('^'..names.close..'$', close)

	local function clear(event)
		event.element.parent[names.scroll][names.textbox].text = ""
	end
	Gui.on_click('^'..names.clear..'$', clear)
end})

return true
```

It's particularly annoying that we are more-or-less forced to decide at module-load time if we want to require all those globals or not (I removed that functionality above).  I can't think of an elegant way to conditionalize it... maybe just look for some magic global variable or something. Or maybe stdlib has some facility for out-of-band configuration data that could be leveraged?  Anyhow, you know everything I do about how to fix this now, good luck :P

**Comment by Nexela** [comment]
Now that I think about it..........   I should just put everything into an STDLIB global instead of polluting everything!

Ps Console works
local console = require("stdlib/path/to/console/cause/i/forgot")
remote.add_interface("my-interface", {console = console})

**Comment by Nexela** [comment]
I have removed the automatic global loading for the next data library push  I will add it as a core function

**Comment by gmt** [comment]
> Now that I think about it.......... I should just put everything into an STDLIB global instead of polluting everything!

That's what I wanted the lazy loading for.  I already implemented it...  It was awesome, worked like a charm.  Until I tried to use it and discovered it was useless in practice :(  Maybe there is some way I can outsmart them using coroutines or something but I'm not super-optimistic; it's likely checked in C++, somewhere.  Now that I have a console to  tool around in, it should be much easier to figure such things out :)

> Ps Console works
> local console = require("stdlib/path/to/console/cause/i/forgot")
> remote.add_interface("my-interface", {console = console})

Really?  For some reason I developed a quasi-superstitious belief that the global loading prevented this from working.  Anyhow sounds like you are several steps ahead of me on this (not surprising) so I'll leave it in your capable hands.  Thanks for all the hard work on stdlib, btw, it's been quite handy.  Feel free to close my ticket.  I'm a happy camper just knowing you're on the case.
